File: ICPC/Practica_para_Gran_Premio_de_Mexico_2/N_Lucky_Number.cpp
#include<bits/stdc++.h>
using namespace std;

#define sz(a)   (int) a.size()
#define all(a)  a.begin(), a.end()

typedef long long ll;
typedef pair<int, int> pi;

int main () {
    ios::sync_with_stdio(false);
    cin.tie(0);

    int n;
    cin >> n;

    vector<int> a(5);
    for (int i = 0; i < n; i++) {
        int x;
        cin >> x;

        a[x % 5]++;
    }

    int res = a[0], mn = min(a[1], a[4]);
    a[1] -= mn, a[4] -= mn;
    res += mn;

    mn = min(a[2], a[3]);
    a[2] -= mn, a[3] -= mn;
    res += mn;

    function<int(int, int, int)> get = [&] (int a, int b, int op) -> int {
        if (op <= 1) {
            int mn = min(a, b / 2);
            int r1 = get(a - mn, b - mn * 2, op + 1) + mn;
            mn = min(a / 3, b);
            int r2 = get(a - mn * 3, b - mn, op + 1) + mn;
            return max(r1, r2);
        }
        
        return a / 5 + b / 5;
    };

    res += max ({
        get(a[2], a[4], 0),
        get(a[1], a[2], 0),
        get(a[3], a[1], 0),
        get(a[4], a[3], 0)
    });

    cout << res << "\n";

    return 0;
}

File: ICPC/Practica_para_Gran_Premio_de_Mexico_2/N_Lucky_Number.py
import sys

def get(a, b, op):
    if op <= 1:
        mn = min(a, b // 2)
        r1 = get(a - mn, b - mn * 2, op + 1) + mn
        mn = min(a // 3, b)
        r2 = get(a - mn * 3, b - mn, op + 1) + mn
        return max(r1, r2)
    
    return a // 5 + b // 5

input = sys.stdin.read

d = input().strip().split()
n = int(d[0])
a = [0] * 5

for i in range(1, n + 1):
    x = int(d[i])
    a[x % 5] += 1

res = a[0]
mn = min(a[1], a[4])
a[1] -= mn
a[4] -= mn
res += mn

mn = min(a[2], a[3])
a[2] -= mn
a[3] -= mn
res += mn

res += max(
    get(a[2], a[4], 0),
    get(a[1], a[2], 0),
    get(a[3], a[1], 0),
    get(a[4], a[3], 0)
)

print(res)
